File: leetcode/981. Time Based Key-Value Store/main.go
package main

type Value struct {
	timestamp int
	value     string
}

type TimeMap struct {
	m map[string][]Value
}

func Constructor() TimeMap {
	return TimeMap{
		m: make(map[string][]Value),
	}
}

func (this *TimeMap) Set(key string, value string, timestamp int) {
	this.m[key] = append(this.m[key], Value{timestamp, value})
}

func (this *TimeMap) Get(key string, timestamp int) string {
	values, ok := this.m[key]
	if !ok {
		return ""
	}

	result := ""

	low, high := 0, len(values)-1

	for low <= high {
		mid := low + (high-low)/2

		if values[mid].timestamp == timestamp {
			return values[mid].value
		} else if values[mid].timestamp < timestamp {
			result = values[mid].value
			low = mid + 1
		} else {
			high = mid - 1
		}
	}

	return result
}

/**
 * Your TimeMap object will be instantiated and called as such:
 * obj := Constructor();
 * obj.Set(key,value,timestamp);
 * param_2 := obj.Get(key,timestamp);
 */

File: leetcode/981. Time Based Key-Value Store/main.rs
use std::collections::HashMap; 

struct TimeMap {
    map: HashMap<String, Vec<(i32, String)>>,
}


/** 
 * `&self` means the method takes an immutable reference.
 * If you need a mutable reference, change it to `&mut self` instead.
 */
impl TimeMap {

    fn new() -> Self {
        TimeMap {
            map: HashMap::new(),
        }
    }
    
    fn set(&mut self, key: String, value: String, timestamp: i32) {
        self.map.entry(key).or_insert(Vec::new()).push((timestamp, value));
    }
    
    fn get(&self, key: String, timestamp: i32) -> String {
        match self.map.get(&key) {
            Some(values) => {
                let mut left = 0;
                let mut right = values.len() as i32 - 1;

                let mut result = "";

                while left <= right {
                    let mid = left + (right - left) / 2;

                    let (mid_ts, mid_val) = &values[mid as usize];

                    if *mid_ts == timestamp {
                        return mid_val.clone();
                    } else if *mid_ts < timestamp {
                        result = mid_val;
                        left = mid + 1;
                    } else {
                        right = mid - 1;
                    }
                }

                result.to_string()
            },
            None => String::new(),
        }
    }
}

/**
 * Your TimeMap object will be instantiated and called as such:
 * let obj = TimeMap::new();
 * obj.set(key, value, timestamp);
 * let ret_2: String = obj.get(key, timestamp);
 */
